### ISSUE TITLE
[v0.4.1] ykman info: Error: Failed connecting to the YubiKey.

### ISSUE THREAD

**Issue by Tormen** [Yubico/yubikey-manager#20]
On archlinux:
Linux horizon 4.12.8-2-ARCH #1 SMP PREEMPT Fri Aug 18 14:08:02 UTC 2017 x86_64 GNU/Linux

YubiKey Manager (ykman) version: 0.4.1
Libraries:
    libykpers 1.18.0
    libu2f-host 1.1.3
    libusb 1.0.21

I am trying "ykman info" on *two* YukiKey version 4 (yk4 and yk4c) and get for both: 
   Error: Failed connecting to the YubiKey.

But heading to https://demo.yubico.com and trying the single-factor authentication works fine.

What am I missing ? ... Maybe because ykman is wip ? Or this is a bug ?

Btw "ykman slot info" seems to work fine:
YubiKey 4
Slot 1: programmed
Slot 2: empty

Also yubikey-personalization-gui recognizes the yubikey correctly:
Firmware Version 4.2.8

**Comment by Tormen** [comment]
"ykman mode" fails as well:

% ykman mode
Usage: ykman [OPTIONS] COMMAND [ARGS]...

Error: Failed connecting to the YubiKey.

% ykman mode o+u+c --touch-eject --autoeject-timeout 15 --chalresp-timeout 15 -f
Usage: ykman [OPTIONS] COMMAND [ARGS]...

Error: Failed connecting to the YubiKey.

**Comment by Tormen** [comment]
Same goes for: 

ykman slot calculate 1 4a6fbb
Usage: ykman [OPTIONS] COMMAND [ARGS]...

Error: Failed to calculate challenge.

((( and also for: ykchalresp -v -N Test
Firmware version 4.2.8
Sending 4 bytes HMAC challenge to slot 1
Yubikey core error: timeout
)))

**Comment by dagheyman** [comment]
The YubiKey 4 have three USB connection modes, OTP, CCID and U2F, all enabled by default. The demo site, yubikey-personalization-gui and `$ ykman slot info` uses OTP, which seems to work fine. It looks like you are having problems with the CCID mode, similar to issue #19. Could you try installing the package `ccid` ?

**Comment by Tormen** [comment]
@Dag: Thanks! - I love any explanation I can get to further understand
my yk4 and will enrich the archlinux yubikey wiki with all info I can get !!

I am currently trying to understand all possible INPUTS and OUTPUTS of
my yk4 and the modes play obviously a deciding role in that :)

Concerning the 3 connection modes:

1. U2F:
When does the yubikey switch to mode U2F ? From what I understand: U2F
is always possible with every YubiKey. It does not need to be
configured. It is never associated to any slot. It is always available
and cannot be deactivated (if your yubikey hardware supports U2F of
course). Correct ?
And it gets activated upon a certain request send to the yubikey while
being inserted, correct ?

2. CCID:
Is it correct to state: Whenever I enable mode CCID (with the
corresponding mode bit), the yubikey will be in CCID mode when inserted. ?
Or would there be for instance the possibility to have the CCID mode
enable only after I long press (activate slot 2, which could possibly be
associated to the CCID mode)?
The other way round: Is it possible to associate CCID to a slot (1 or 2)
?? or is it never associated to any slot. But CCID is rather
activated/deactivated by setting the CCID mode.

And I think I know that I would like mode 6, but am not sure if I want
mode 86, so I don't know if I want to use ykman mode --touch-eject or not.
I find your explanation rather unhelpful. To give you an idea of what I
mean: You write """the button toggles the state of the smartcard between
ejected and inserted"""
... and this leaves me with many questions. Some of these questions are:
(H.1) So, *everytime* I press on the yubikey I insert/eject the smartcard ??
(H.2) No matter if I press and hold 3-4 seconds (slot 2) or shorter
(slot 1) ??
(H.3) No matter how the slot is configured ??).
--> Could you please walk me to a use-case with and without touch-eject
of the yubikey as smartcard ?

3. OTP:
Is that correct ? : The OTP mode must be configured to slot 1 and/or 2.
If done so, it can then be activated after insertion of the yk by either
a short or a long press on the yk button. Which will ask the device to
trigger a OTP.
At the same time a Yubikey-OTP programmed slot allows also *always* to
do a Yubikey Challenge-Response using the AES secret key ?
Is it possible to configure a slot to ONLY Yubikey OTP or ONLY Yubikey
Challenge-Response?
And what about the HMAC Challenge-Response? Is there also a link to OTP
(maybe the OATH-HOTP ?) Or can it only exist as a separate slot
configuration ?

If you could provide me with answers to my above questions, I promise
that I will make them count (!) by putting the info in comprehensive
form into the Archlinux wikipage on the Yubikey. Maybe creating a new
page for yk4 actually, because some of the answers depend on the
hardware capabilities of your yk ...
I know that these questions go a bit above the topic of this ticket, but
I am desperately seeking for answers. And there seems no comprehensive
answers to these questions. 

If you want, I could also open a new support ticket: "Understanding the
modes" ?

Otherwise: The package ccid was and is already installed :)

**Comment by dagheyman** [comment]
Ok, first for the troubleshooting of the actual issue. Does installing `python-pyotherside` or `yubico-yubioath-desktop` change anything as suggested in #19 ? These steps might be worth trying as well: https://github.com/Yubico/yubioath-desktop/issues/172#issuecomment-321343125

----

Regarding your questions, I'll try my best to answer them.

> When does the yubikey switch to mode U2F ? From what I understand: U2F
is always possible with every YubiKey. It does not need to be
configured. It is never associated to any slot. It is always available
and cannot be deactivated (if your yubikey hardware supports U2F of
course). Correct ?

U2F functionality is enabled by default in YK4s and currently sold NEOs. You are correct in that it doesn't need any configuration and doesn't effect the two slots. The functionality can be disabled, by disabling the U2F mode. U2F as a transport is also used in this tool for mode switching, when it is the only transport mode available.

Example:
`$ ykman mode u`
will disable OTP and CCID, and only leave U2F enabled.
`$ ykman mode u+o+c`
will then enable all modes again, using the U2F transport.

> CCID:
Is it correct to state: Whenever I enable mode CCID (with the
corresponding mode bit), the yubikey will be in CCID mode when inserted. ?
Or would there be for instance the possibility to have the CCID mode
enable only after I long press (activate slot 2, which could possibly be
associated to the CCID mode)?
The other way round: Is it possible to associate CCID to a slot (1 or 2)
?? or is it never associated to any slot. But CCID is rather
activated/deactivated by setting the CCID mode.

If CCID mode is enabled (the modes are not mutually exclusive, they are usually all enabled), this tool will prefer it as a transport mode. That is probably why you are failing to connect to the key. It is not related to the two slots (short press or long press) in any way, it functions as a smart card reader with a smart card inserted in it.  The `--touch-eject` flag can only be used when the device is in CCID mode _only_, and I wouldn't recommend it because it would prevent you from using the U2F and OTP features of the YubiKey.

> OTP:
Is that correct ? : The OTP mode must be configured to slot 1 and/or 2.
If done so, it can then be activated after insertion of the yk by either
a short or a long press on the yk button. Which will ask the device to
trigger a OTP.
At the same time a Yubikey-OTP programmed slot allows also always to
do a Yubikey Challenge-Response using the AES secret key ?
Is it possible to configure a slot to ONLY Yubikey OTP or ONLY Yubikey
Challenge-Response?
And what about the HMAC Challenge-Response? Is there also a link to OTP
(maybe the OATH-HOTP ?) Or can it only exist as a separate slot
configuration ?

So the term OTP here can refer to (at least) two things. If the transport mode OTP is enabled, the two YubiKey Slots, long press and short press, can be configured and used. These slots can have one of the following credentials configured: a [Yubico OTP](https://developers.yubico.com/OTP/) (which is what comes preconfigured in the short press slot on a new key), a static password, a challenge-response credential, a OATH-HOTP credential. All this functionality is found in the `$ ykman slot` commands.

I hope I clarified some stuff. Thanks for your questions, they expose in what areas documentation is lacking, even though as you mention I think support is a better channel for these more general questions.

**Comment by Tormen** [comment]
About the issue at hand: 

Hah, That the CCID transport mode is used by ykman is is important to know!

Now that I started pscd is running, it works like charme!

So IMHO it would make sense, that ykman gives some hints at this point (ykman "knows" that it is trying to use the CCID connection and could inform the user, because I don't find that intuitve at all ;) ... and could propose to the user to check if the pcscd is running).

---------------------------------------------------

Cool. THANKS A LOOOOOOOOOT for your answers !!!

So cool, when you finally understand your little friend yk4 :)

SUPPORT: I actually send a long list of questions like the above to support. The answer: They don't have the ressources to treat such questions :/
Even though I had offered them to put their answers in the archlinux wiki and/or compile the missing man page for ykman.

I delivered my end of the bargain and put all the info I got so far from Yubico in this page. I hope you like the structure and tone and maybe you could help fill in some of the "TODO" / "..." ? Ideally the document as is should be useful for Yubico (as missing manual for the YubiKey 4) or as missing man-page for ykman maybe ?? 

https://wiki.archlinux.org/index.php/Yubikey

Please let me know.

**Comment by Tormen** [comment]
2 Questions:
* How can I figure out if I set the --touch-eject flag is set or not?
* Is there a way to read/list all flags from the yk4 ?

**Comment by dagheyman** [comment]
Great the the issue is resolved! I'll look into providing better error messages as you suggested, and improving the official documentation. I'm glad the Arch wiki is there, but I think it's better if it just keeps being maintained by the Arch community.

---


> How can I figure out if I set the --touch-eject flag is set or not? (ykman mode does not seem to tell me) Same for the two timeouts that can be set via ykman mode: How can I read the current settings ?

No, these flags can not be read from the device once written. However, the behaviour of the device should change when `--touch-eject` is set, you should be able to eject/insert the smartcard by pressing the button. The LED should indicate if the card is inserted or not as well.

> Can I have Yubico OTP configured on slot 1 AND be able to use Challenge Response (yubico mode) with this slot 1 ?? (ykman slot calculate 1 4a6fbb fails for me)
Or will I need a second slot to have Yubico OTP /and/ Challenge Response (ykchalresp) ??

A slot has _either_ a Yubico OTP or a challenge-response credential configured. So a Yubico OTP in slot 1 and a challenge response secret in slot 2 should work fine.

> More general:
You wrote: """These slots can have one of the following credentials configured: a Yubico OTP (which is what comes preconfigured in the short press slot on a new key), a static password, a challenge-response credential, a OATH-HOTP credential. """ Is it "you can have one AND ONLY ONE" or "one or multiple" of the following configured (on a single slot) ? - So can I combine any of these ? As OTP, OATH-HOTP and Static password are all TOUCH triggered (Challenge response credential cannot be triggered through button touch, or??), they are probably mutually exclusive. But as Challenge-Response is triggered via API calls, this could potentially be combined, no ?

One, and only one, type of credential per slot. A challenge-response credential may require touch, see `$ ykman slot chalresp -h`.

> The LED of my key is shortly flashing every two-three seconds; constantly. Is that because CCID mode is activated? What the is the benefit of this blinking? Can I disable or suppress the LED (without the CCID mode) under Linux?

I believe the best current workaround to get the LED to blink less is to disable CCID.

> The more I learn about yk4 I think ykman is well thought and it's command structure makes sense and also will help users to better understand all the possibilities.

Good to hear! Feedback is always welcome.

> How can this be? - I mean the first "nothing to do..." is possible, even though I thought I had set my yubikey to mode 86 the last time. But then the ykman mode 86 shoudl have something to do (changing the touch-eject flag), no ??

The only valid modes when using numbers is 0 - 6 (see https://github.com/Yubico/yubikey-manager/blob/master/ykman/util.py#L94). The extra flags are not part of the mode in that sense, they just need to be set at the same time as the mode is set. 86 is not a valid mode. It might be a bug that the tool accepts higher values.

**Comment by Tormen** [comment]
@dagheyman : Thanks a lot! I put in all your input into the Archwiki page. For now it only seems maintained by ... me. Maybe a YubiKey wikipage would be nice allowing change-propositions that you could review and accept/accept in modified version or reject?!

Just to benefit from the COMMUNITY in building and maintaining an extensive documentation. That to my knowledge does NOT exist. The YubiKey manual does not talk about yk4 I think and I already found several inconsistancies...

In the archlinux wiki are still several TODOs left. It would be great to get them sorted out :)

**Comment by christianreiser** [comment]
running the manager with sudo fixed the problem for me on Ubuntu:
`sudo ./yubikey-manager-qt-1.1.5-linux.AppImage`
